wpimath/src/main/java/edu/wpi/first/math/geometry/Quaternion.java

```diff
@@ -126,14 +187,110 @@ public Quaternion inverse() {
     if (norm == 0.0) {
       return new Quaternion();
     } else {
       return new Quaternion(getW() / norm, getX() / norm, getY() / norm, getZ() / norm);
     }
   }
 
+  /**
+   * Rational power of a quaternion.
+   *
+   * @param t the power to raise this quaternion to.
+   * @return The quaternion power
+   */
+  public Quaternion pow(double t) {
+    // q^t = e^(ln(q^t)) = e^(t * ln(q))
+    return this.log().times(t).exp();
+  }
+
+  /**
+   * Matrix exponential of a quaternion.
+   *
+   * @param adjustment the "Twist" that will be applied to this quaternion.
+   * @return The quaternion product of exp(adjustment) * this
+   */
+  public Quaternion exp(Quaternion adjustment) {
+    return adjustment.exp().times(this);
+  }
+
+  /**
+   * Matrix exponential of a quaternion.
+   *
+   * <p>source: https://en.wikipedia.org/wiki/Quaternion#Exponential,_logarithm,_and_power_functions
+   *
+   * @return The Matrix exponential of this quaternion.
+   */
+  public Quaternion exp() {
+    // q = s(scalar) + v(vector)
+    // exp(s)
+    var scalar = Math.exp(m_w);
+
+    // ||v||
+    var axial_magnitude = Math.sqrt(getX() * getX() + getY() * getY() + getZ() * getZ());
+    // cos(||v||)
+    var cosine = Math.cos(axial_magnitude);
+
+    double axial_scalar;
+
+    if (axial_magnitude < 1e-9) {
+      // Taylor series of sin(x)/x near x=0: 1 − x²/6 + x⁴/120 + O(n⁶)
+      var axial_magnitude_sq = axial_magnitude * axial_magnitude;
+      var axial_magnitude_sq_sq = axial_magnitude_sq * axial_magnitude_sq;
+      axial_scalar = 1 - axial_magnitude_sq / 6 + axial_magnitude_sq_sq / 120;
+    } else {
+      axial_scalar = Math.sin(axial_magnitude) / axial_magnitude;
+    }
+
+    // exp(s) * (cos(||v||) + v * sin(||v||) / ||v||)
+    return new Quaternion(
+        cosine * scalar,
+        getX() * axial_scalar * scalar,
+        getY() * axial_scalar * scalar,
+        getZ() * axial_scalar * scalar);
+  }
+
+  /**
+   * Log operator of a quaternion.
+   *
+   * @param end The quaternion to map this quaternion onto.
+   * @return The "Twist" that maps this quaternion to the argument.
+   */
+  public Quaternion log(Quaternion end) {
+    return end.times(this.inverse()).log();
+  }
+
+  /**
+   * The Log operator of a general quaternion.
+   *
+   * <p>source: https://github.com/moble/quaternionic/blob/main/quaternionic/algebra.py#L233
+   *
+   * <p>For unit quaternions, this is equivalent to @see Quaternion#toRotationVector().
+   *
+   * @return The logarithm of this quaternion.
+   */
+  public Quaternion log() {
+    var vnorm = Math.sqrt(getX() * getX() + getY() * getY() + getZ() * getZ());
+    if (vnorm <= 1e-9 * Math.abs(getW())) {
+      if (getW() < 0) {
+        if (Math.abs(getW() + 1) > 1e-9) {
+          return new Quaternion(Math.log(-getW()), Math.PI, 0, 0);
+        } else {
+          return new Quaternion(0, Math.PI, 0, 0);
+        }
+      } else {
+        return new Quaternion(Math.log(getW()), 0, 0, 0);
+      }
```

Comment:
I've re-implemented this w/ passing all the same tests using a similar derivation to `toRotationVector`. It passes all the same tests that the previous implementation passed, with a specific but non-functional change for the first edge case.

Reply:
We should probably still have tests for the edge cases- The logic checks out, but if reviews were infallible, we wouldn't need tests in the first place. I'm also a bit iffy about using a Taylor series approximation for when v_norm is close to 0 instead of the previous handling since it feels a lot less direct (and more likely for errors to accumulate).